### PR TITLE
Use virtual paths to specify exact mod to enable

### LINKED ISSUE #4183: Mod conflicts if same mod in many modpacks
I have two modpacks, my own one `etherial_modpack` (yes, it has modpack.txt) and dreambuilder_modpack, they have some similar mods inside of them both, like quartz, bedrock, signs, etc

Now I've created my world, in settings I'm enabling my own modpack ONLY (double click on it), save, trying to run and I see this error:

```
2016-06-02 21:32:28: WARNING[Main]: Mod name conflict detected: "bedrock"
2016-06-02 21:32:28: WARNING[Main]: Will not load: D:\Newprog\minetest\bin\..\mods\dreambuilder_modpack\bedrock
2016-06-02 21:32:28: WARNING[Main]: Will not load: D:\Newprog\minetest\bin\..\mods\etherial_modpack\bedrock
... etc
```

In world.mt I see lots of completely wrong load_mod_ entries with entries I have not selected, like mesecons, I don't have mesecons in my modpack. 

This is irritating and confusing.

Minetest 0.4.14-5699980-5699980-dirty (Windows)
Using Irrlicht 1.8.1
Build info: VER=0.4.14-5699980-5699980-dirty BUILD_TYPE=Release RUN_IN_PLACE=1 USE_GETTEXT=1 USE_SOUND=1 USE_CURL=1 USE_FREETYPE=1 USE_LUAJIT=1 STATIC_SHAREDIR="."

### PR DESCRIPTION
**Pull request by rubenwardy**
Currently, world.mt uses true or false to enable a mod based on its name. When there are multiple mods with the same name, Minetest will fail to load. This isn't very user friendly or obvious at all

This PR uses virtual paths like `mods/modpack1/moddir` to enable the mod instead of `true`. This allows exact copies of the mod to be enabled

```
load_mod_mymod = mods/modpack1/moddir
```

Thanks to pgimeno for paving the way for this

Fixes #4183

## To do

- [x] Implement UI
- [x] share/ and env/ support
- [x] Migrate mod if no paths found

## How to test

Here are 3 mods in 2 modpacks, all with the same name

[load_mod_test.zip](https://github.com/minetest/minetest/files/7576805/load_mod_test.zip)

### CODE REVIEW

File: src/content/mods.cpp [hunk 1]
```diff
@@ -293,29 +298,38 @@ void ModConfiguration::addMods(const std::vector<ModSpec> &new_mods)
...
+			load_mod_names[name.substr(9)] = value;
 	}
 
 	std::vector<ModSpec> addon_mods;
-	for (const std::string &i : mods) {
-		std::vector<ModSpec> addon_mods_in_path = flattenMods(getModsInPath(i));
+	std::unordered_map<std::string, std::vector<std::string>> candidates;
+
+	for (const std::string &modPath : modPaths) {
+		std::vector<ModSpec> addon_mods_in_path = flattenMods(getModsInPath(modPath, "mods"));
```
Comment (rubenwardy) [inline src/content/mods.cpp:319]: To do: share mods should use `share/` rather than `mods/`

Comment (rubenwardy) [reply src/content/mods.cpp:319]: OK, so the mainmenu doesn't support `share/mods`, but does support an environment variable for loading mods

**Comment by rubenwardy** [comment]
One potential UX shortfall here: if the virtual path doesn't exist, the mod just won't load - the game itself will load though.
This would make naively sharing worlds hard, as different players may have mods installed to different locations

Solutions:

1. Do nothing
2. Throw an error rather than continuing to load the game, let the player sort it out (to disable deleted mods, you would now need to open Config Mods and then click Save, which isn't very intuitive)
3. Automatically select other mods with the same name when loading the game (errr.... potentially dodgy)
4. Automatically select other mods with the same name when opening config mods (probably better, if there's a warning)

**Comment by SmallJoker** [comment]
How about "4", but only do the automatic selection if there's exactly one mod with the matching name?

**Comment by rubenwardy** [comment]
> How about "4", but only do the automatic selection if there's exactly one mod with the matching name? Otherwise show a warning.

Done - it now automatically selects a mod in a different path if there's only one mod with that name present. I haven't handled the case with multiple mods because I'd like to improve the usability of the config mods dialog in the future, and don't want to make this PR much more complicated

**Comment by SmallJoker** [comment]
Breaks CSMs:
```
2022-01-02 20:58:43: WARNING[Main]: Mod name conflict detected: "colored_names"
2022-01-02 20:58:43: WARNING[Main]: Will not load: /data/Minetest/run/bin/../clientmods/colored_names
2022-01-02 20:58:43: WARNING[Main]: Will not load: /data/Minetest/run/bin/../clientmods/colored_names
2022-01-02 20:58:43: WARNING[Main]: Mod name conflict detected: "inspector"
2022-01-02 20:58:43: WARNING[Main]: Will not load: /data/Minetest/run/bin/../clientmods/inspector
2022-01-02 20:58:43: WARNING[Main]: Will not load: /data/Minetest/run/bin/../clientmods/inspector
```
RUN_IN_PLACE=1

**Comment by rubenwardy** [comment]
I blissfully forgot that CSMs were a thing

File: builtin/mainmenu/dlg_config_world.lua [hunk 2]
```diff
@@ -205,14 +205,19 @@ local function handle_buttons(this, fields)
...
 			if not mod.is_modpack and
 					not mod.is_game_content then
 				if modname_valid(mod.name) then
-					worldfile:set("load_mod_" .. mod.name,
-						mod.enabled and "true" or "false")
+					if mod.enabled then
+						worldfile:set("load_mod_" .. mod.name, mod.virtual_path)
+						was_set[mod.name] = true
+					elseif not was_set[mod.name] then
+						worldfile:set("load_mod_" .. mod.name, "false")
```
Comment (Desour) [inline builtin/mainmenu/dlg_config_world.lua:219]: Can't we just not write these "false" lines?
They leak the names of all installed mods, which makes sending worlds to other persons a bit awkward.
And afaik their only real use was to provide defaults to edit if you want to enable mods without a gui.

Comment (rubenwardy) [reply builtin/mainmenu/dlg_config_world.lua:219]: This is already done before this PR. This should be another issue and another PR

**Comment by SmallJoker** [comment]
grafik
The main menu is broken. It lists all mods as enabled. Workaround: enable all, disable all. The PR seems to work well other than in this aspect.

**Comment by rubenwardy** [comment]
>  The main menu is broken. It lists all mods as enabled. Workaround: enable all, disable all. The PR seems to work well other than in this aspect.

Fixed, was missing a check on the value in world.mt

**Review by SmallJoker — APPROVED** [review]
Works-

**Pull request merged**